File: code_snippets/api-guide/others/api-rate-limits.py
import requests
import time


def make_request(url, headers):
    try:
        response = requests.get(url, headers=headers)
        response.raise_for_status()
        return response.json()
    except requests.exceptions.HTTPError as err:
        error_response = err.response.json()
        if (
            "status" in error_response
            and error_response["status"]["code"] == "CONN_THROTTLED"
        ):
            print("Rate limit exceeded. Waiting for 15 seconds before retrying...")
            time.sleep(15)
            return make_request(url, headers)  # Retry the request after waiting
        else:
            # Handle other types of errors
            print("Error:", err)
            return None


url = "https://api.clarifai.com/v2/models/types"
headers = {"Authorization": "Key YOUR_PAT_HERE"}

response_data = make_request(url, headers)

if response_data:
    print(response_data)

File: code_snippets/api-guide/others/api-rate-limits.js
const axios = require('axios');
async function makeRequest(url, headers) {
    try {
        const response = await axios.get(url, { headers });
        return response.data;
    } catch (error) {
        if (error.response && error.response.data.status && error.response.data.status.code === 'CONN_THROTTLED') {
            console.log("Rate limit exceeded. Waiting for 15 seconds before retrying...");
            await delay(15000); // Wait for 15 seconds before retrying
            return makeRequest(url, headers); // Retry the request after waiting
        } else {
            // Handle other types of errors
            console.error("Error:", error);
            return null;
        }
    }
}
// Helper function to delay execution
function delay(ms) {
    return new Promise(resolve => setTimeout(resolve, ms));
}
const url = "https://api.clarifai.com/v2/models/types";
const headers = { 'Authorization': 'Key YOUR_PAT_HERE' };
makeRequest(url, headers)
    .then(responseData => {
        if (responseData) {
            console.log(responseData);
        }
    });
